File: tutorial/python/0-IVF.py
# -*- coding: utf-8 -*-

"""
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
"""

import numpy as np
import faiss
import time

# 维度
d = 64
# 数据库大小
nb = 100000
# 查询向量数量
nq = 10000

# --- 1. 生成数据 ---
print("Generating fake data...")
# 设置随机种子以保证结果可复现
np.random.seed(1234)

# 创建数据库向量 xb
xb = np.random.random((nb, d)).astype('float32')
# 对数据进行一些轻微的、非均匀的调整，使其不完全是随机分布
xb[:, 0] += np.arange(nb) / 1000.

# 创建查询向量 xq
xq = np.random.random((nq, d)).astype('float32')
# 同样对查询向量进行调整
xq[:, 0] += np.arange(nq) / 1000.

print("Data generated.")

# --- 2. 构建并训练索引 ---
nlist = 100  # 将数据库分为100个倒排列表
k = 4        # 我们想要查找最近的4个邻居

# 量化器，这里使用精确的L2距离索引
quantizer = faiss.IndexFlatL2(d)
# 创建IVFFlat索引
index = faiss.IndexIVFFlat(quantizer, d, nlist)

# 训练索引
assert not index.is_trained
print("Training index...")
index.train(xb)
assert index.is_trained
print("Training complete.")

# 添加向量到索引
print("Adding vectors to index...")
index.add(xb)
print(f"Index contains {index.ntotal} vectors.")


# --- 3. 设置并行模式并准备延迟统计 ---

# 1. 设置为 parallel_mode = 0 (在IVF搜索中禁用查询间的并行)
#    这使得我们可以更清晰地测量单个查询的延迟
index.parallel_mode = 0

# 2. 将 OMP 线程数设为 1，以精确测量单线程下的真实挂钟延迟
#    这模拟了典型的 Web 服务中单个工作进程的行为
faiss.omp_set_num_threads(1)

print(f"Starting search with nq={nq}, k={k}, nlist={nlist}, "
      f"parallel_mode={index.parallel_mode}, OMP_threads={faiss.omp_get_max_threads()}")


# --- 4. 逐个执行查询并收集延迟数据 ---
latencies_ms = []
print("Measuring per-query latency...")

# 为了避免首次查询的“冷启动”开销影响统计，可以先进行一次预热查询
index.search(xq[:1], k)

for i in range(nq):
    # 获取单个查询向量，并确保其为 2D 数组
    query_vector = xq[i:i+1]

    # 记录开始时间
    start_time = time.perf_counter()
    # 执行搜索
    D, I = index.search(query_vector, k)
    # 记录结束时间
    end_time = time.perf_counter()

    # 计算耗时并转换为毫秒，然后添加到列表中
    latency = (end_time - start_time) * 1000
    latencies_ms.append(latency)

print("Latency measurement complete.")

# --- 5. 分析并打印延迟统计结果 ---
latencies_np = np.array(latencies_ms)

avg_latency = np.mean(latencies_np)
min_latency = np.min(latencies_np)
max_latency = np.max(latencies_np)
p50_latency = np.median(latencies_np) # P50 就是中位数
p99_latency = np.percentile(latencies_np, 99)

print("\n--- Latency Stats (ms) ---")
print(f"Average : {avg_latency:.4f} ms")
print(f"Min     : {min_latency:.4f} ms")
print(f"Max     : {max_latency:.4f} ms")
print(f"P50 (Median) : {p50_latency:.4f} ms")
print(f"P99     : {p99_latency:.4f} ms")
print("--------------------------\n")

# 注意：标准的 Faiss Python API 无法直接提供 C++ 版本中
# quantization_us 和 list_scan_us 这样的内部耗时分解。
# 我们这里测量的是每个 search() 调用的总延迟。


# --- 6. (可选) 批量搜索并验证结果正确性 ---
print("Performing batch search for verification...")
# 为了验证结果，我们进行一次完整的批量搜索
D_batch, I_batch = index.search(xq, k)

# 打印最后5个查询的结果，与 C++ 示例的输出格式保持一致
print("I (last 5 results)=")
for i in range(nq - 5, nq):
    for j in range(k):
        print(f"{I_batch[i][j]:5d}", end=" ")
    print()

File: tutorial/cpp/0-IVF.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <random>
#include <vector>
#include <numeric>
#include <algorithm>
#include <omp.h>

#include <faiss/IndexFlat.h>
#include <faiss/IndexIVFFlat.h>

using faiss::QueryLatencyStats;
using idx_t = faiss::idx_t;

// === 新增一个结构体来保存单次运行的结果 ===
struct BenchmarkResult {
    double total_wall_time_s = 0.0;
    double qps = 0.0;
    double avg_latency_ms = 0.0;
    double p50_latency_ms = 0.0;
    double p99_latency_ms = 0.0;
};

int main() {
    int d = 64;      // dimension
    int nb = 100000; // database size
    int nq = 10000;  // nb of queries

    printf("Generating data...\n");
    std::mt19937 rng;
    std::uniform_real_distribution<> distrib;

    // 使用 std::vector 替代裸指针，更安全
    std::vector<float> xb(d * nb);
    std::vector<float> xq(d * nq);

    for (int i = 0; i < nb; i++) {
        for (int j = 0; j < d; j++)
            xb[d * i + j] = distrib(rng);
        xb[d * i] += i / 1000.;
    }

    for (int i = 0; i < nq; i++) {
        for (int j = 0; j < d; j++)
            xq[d * i + j] = distrib(rng);
        xq[d * i] += i / 1000.;
    }

    int nlist = 100;
    int k = 4;

    printf("Training and adding vectors to index...\n");
    faiss::IndexFlatL2 quantizer(d);
    faiss::IndexIVFFlat index(&quantizer, d, nlist);
    index.train(nb, xb.data());
    index.add(nb, xb.data());
    index.parallel_mode = 0;

    // === 基准测试循环设置 ===
    std::vector<int> thread_counts = {1, 10, 20, 30, 40, 50, 60};
    const int num_runs_per_thread_setting = 3;
    
    // 用于存储最终平均结果的向量
    std::vector<std::pair<int, BenchmarkResult>> final_results;

    printf("Starting benchmark...\n");

    // === 外层循环：遍历不同的线程数设置 ===
    for (int threads : thread_counts) {
        omp_set_num_threads(threads);
        
        std::vector<BenchmarkResult> run_results;
        printf("\n--- Testing with %d OMP threads ---\n", threads);

        // === 内层循环：为每个线程数设置重复运行多次 ===
        for (int i = 0; i < num_runs_per_thread_setting; ++i) {
            idx_t* I = new idx_t[k * nq];
            float* D = new float[k * nq];
            auto latency_stats = new QueryLatencyStats[nq];

            double t_start = omp_get_wtime();
            
            // 修正函数名: search_stats -> search_with_stats
            index.search_stats(nq, xq.data(), k, D, I, nullptr, latency_stats);

            double t_end = omp_get_wtime();
            
            // --- 数据收集 ---
            BenchmarkResult current_run;
            current_run.total_wall_time_s = t_end - t_start;
            current_run.qps = nq / current_run.total_wall_time_s;

            std::vector<double> latencies_ms;
            for (int j = 0; j < nq; ++j) {
                latencies_ms.push_back(latency_stats[j].total_us / 1000.0);
            }
            std::sort(latencies_ms.begin(), latencies_ms.end());
            
            current_run.avg_latency_ms = std::accumulate(latencies_ms.begin(), latencies_ms.end(), 0.0) / nq;
            current_run.p50_latency_ms = latencies_ms[nq / 2];
            current_run.p99_latency_ms = latencies_ms[int(nq * 0.99)];
            
            run_results.push_back(current_run);

            printf("Run %d/%d: Total Time = %.4f s, QPS = %.2f\n", 
                   i + 1, num_runs_per_thread_setting, 
                   current_run.total_wall_time_s, current_run.qps);

            delete[] I;
            delete[] D;
            delete[] latency_stats;
        }

        // === 计算平均值（忽略第一次热身运行） ===
        BenchmarkResult avg_result;
        double sum_total_time = 0.0, sum_qps = 0.0, sum_avg_lat = 0.0, sum_p50 = 0.0, sum_p99 = 0.0;
        
        for (int i = 1; i < num_runs_per_thread_setting; ++i) { // 从 i=1 开始，忽略第一次
            sum_total_time += run_results[i].total_wall_time_s;
            sum_qps += run_results[i].qps;
            sum_avg_lat += run_results[i].avg_latency_ms;
            sum_p50 += run_results[i].p50_latency_ms;
            sum_p99 += run_results[i].p99_latency_ms;
        }

        int count = num_runs_per_thread_setting - 1;
        avg_result.total_wall_time_s = sum_total_time / count;
        avg_result.qps = sum_qps / count;
        avg_result.avg_latency_ms = sum_avg_lat / count;
        avg_result.p50_latency_ms = sum_p50 / count;
        avg_result.p99_latency_ms = sum_p99 / count;

        final_results.push_back({threads, avg_result});
    }

    // === 最后，打印格式化的总结果表格 ===
    printf("\n\n===== Faiss Benchmark Summary (Average of last %d runs) =====\n", num_runs_per_thread_setting - 1);
    printf("--------------------------------------------------------------------------------------------------\n");
    printf("| %-12s | %-20s | %-15s | %-20s | %-15s | %-15s |\n", 
           "OMP Threads", "Total Time (s)", "QPS", "Avg Latency (ms)", "P50 Latency (ms)", "P99 Latency (ms)");
    printf("--------------------------------------------------------------------------------------------------\n");

    for (const auto& result_pair : final_results) {
        printf("| %-12d | %-20.4f | %-15.2f | %-20.4f | %-15.4f | %-15.4f |\n", 
               result_pair.first,
               result_pair.second.total_wall_time_s,
               result_pair.second.qps,
               result_pair.second.avg_latency_ms,
               result_pair.second.p50_latency_ms,
               result_pair.second.p99_latency_ms);
    }
    printf("--------------------------------------------------------------------------------------------------\n");

    return 0;
}